docs/ibc/spec.mdx

```diff
@@ -0,0 +1,138 @@
+
+  * **`OnRecvPacket(sourceClientId, destClientId, sequence, payload)`** – Executed on the **destination chain** to process incoming data. It must return an **Acknowledgement** (success or error).
+  * **`OnAcknowledgePacket(sourceClientId, destClientId, sequence, payload, acknowledgement)`** – Executed on the **source chain** once an acknowledgement is verified, finalizing or compensating the originating action.
+  * **`OnTimeoutPacket(sourceClientId, destClientId, sequence, payload)`** – Executed on the **source chain** if no acknowledgement is received before `timeoutTimestamp`, allowing rollback or refunds.
+
+<Warning>For packets with multiple payloads, execution is **atomic**. If the `OnRecvPacket` callback fails for even one payload, the entire packet operation is considered a failure. Any state changes from other successful callbacks in the same packet **must be reverted**.</Warning>
+
+
+
+## Packet Lifecycle
+
+The end‑to‑end flow from send to completion involves five discrete steps. Expand any step below for details.
+
+<Accordion title="1. SendPacket (Source Chain)">
+The sending application submits one or more `Payload` objects. The IBC handler wraps them in a `Packet`, assigns a new **sequence** number, and stores a cryptographic commitment at the **Packet Commitment** path.
+</Accordion>
+
+<Accordion title="2. RecvPacket (Destination Chain)">
+A relayer delivers the `Packet` together with a proof of its commitment. The destination chain’s IBC handler:
+
+1. Locates the correct light client via `destClientId`.
+2. Verifies the proof.
+3. Stores a **Receipt** at the **Packet Receipt** path.
+4. Invokes the destination module’s `OnRecvPacket` callback.
+
+</Accordion>
+
+<Accordion title="3. WriteAcknowledgement (Destination Chain)">
+The destination application returns an **Acknowledgement** (success data or error) from `OnRecvPacket`. The IBC handler commits this acknowledgement at the **Acknowledgement** path, making it ava>
+
+For applications with long-running logic, this process can be asynchronous. The app can process the packet later and call a separate `WriteAcknowledgement` function to submit the result.
+</Accordion>
+
+<Accordion title="4. AcknowledgePacket (Source Chain)">
+A relayer brings the acknowledgement and its proof back to the source chain. After proof verification, the source chain’s handler:
+
+* Deletes the original packet commitment.
+* Calls `OnAcknowledgePacket` on the sending application to finalize the cross‑chain operation.
+
+</Accordion>
+
+<Accordion title="5. TimeoutPacket (Source Chain)">
+If no acknowledgement arrives before `timeout`, a relayer submits a non‑existence proof for the **Packet Receipt**. The source chain deletes the packet commitment and triggers `OnTimeoutPacket`, allowing the application to revert or refund the original transaction.
```

Comment:
Nit: the word "arrives" here is a little ambiguous. If the no receipt been _written_ before the timeout timestamp, the relayer would do this. It's not dependent on wether or not it arrived back at the source chain or not. Technically the acknowledgment can _never_ arrive, if nobody bothers to relay it - many applications don't actually do anything with a success ack. With ICS20 Transfer for instance, the ack does nothing useful - and we don't really need to do it if the gas cost is very high for instance.